src/events/etcd.py

```
@@ -195,3 +220,60 @@ def update_admin_password(self, admin_secret_id: str) -> None:
                         logger.error(e)
         except (ModelError, SecretNotFoundError) as e:
             logger.error(e)
+
+    def update_private_key(self, private_key_id: str, tls_type: TLSType) -> None:
+        """Update the private key in etcd."""
+        logger.debug("Updating TLS private key.")
+
+        tls_relation = (
+            self.charm.state.peer_tls_relation
+            if tls_type == TLSType.PEER
+            else self.charm.state.client_tls_relation
+        )
+
+        tls_requirer = (
+            self.charm.tls_events.peer_certificate
+            if tls_type == TLSType.PEER
+            else self.charm.tls_events.client_certificate
+        )
+        try:
+            secret_content = get_secret_from_id(self.charm.model, private_key_id).get(
+                "private-key"
+            )
+        except (ModelError, SecretNotFoundError) as e:
+            logger.error(e)
+            return
```

This looks risky. This effectively fails silently. I believe we should let the exception propagate and put the charm in a blocked state when this happens -- mainly `on_config_changed`

---

@reneradoi this would be relevant for updating the admin password as well.

---

I agree with Mehdi, in case of failure there needs to be some feedback to the user. This is more difficult for secrets than for an action. Most sane option (without blocking the workload) is to keep everything running and set the unit status to blocked.

For the admin password, I had prepared a PR for this (see [#12](https://github.com/canonical/charmed-etcd-operator/pull/12)). Mykola asked for holding this back until we have the status implementation figured out. His reasoning was, that the blocking status might get overwritten by something else and the user feedback is gone then.

We could keep the logic here as it is for now and update #12 for the private keys, imo.